stan/math/prim/fun/trace_gen_quad_form.hpp

```diff
@@ -68,10 +60,11 @@ inline return_type_t<TD, TA, TB> trace_gen_quad_form(
  * @throw std::domain_error if A cannot be multiplied by B or B cannot
  * be multiplied by D.
  */
-template <int RD, int CD, int RA, int CA, typename TB, int RB, int CB>
-inline double trace_gen_quad_form(const Eigen::Matrix<double, RD, CD> &D,
-                                  const Eigen::Matrix<double, RA, CA> &A,
-                                  const Eigen::Matrix<double, RB, CB> &B) {
+template <typename TD, typename TA, typename TB,
+          typename = require_all_same_t<double, value_type_t<TD>,
+                                        value_type_t<TA>, value_type_t<TB>>,
+          typename = require_all_eigen_t<TD, TA, TB>>
```

Comment:
`require_all_eigen_vt<std::is_floating_point, TD, TA, TB>`

If we ever worry about mixing types of floats we could also have
```
require_same_same_vt<TD, TA, TB>
```

Reply:
Though the bottom may be better as a static assert to the user

Reply:
Static assert is not what we want here - if that require is not satisfied, other overload must be chosen.

Reply:
Anyway I would not complicate here, since this overload can be removed once transpose is generalized too.